### ISSUE TITLE
FlowScreenComponentsBasePack - Flow Combobox - Error When Selecting a Value from a Record Variable

### ISSUE THREAD

**Issue by Garywoo** [alexed1/LightningFlowComponents#861]
While setting values in a Flow Comobobx component, within the CPE of an action or screen component, if you try and use a value from a record variable (i.e. VarOppoprtunity.Id), the following error occurs:

> Something went wrong
> Save your flow and refresh the page. If the problem persists, contact Salesforce Customer Support. Error ID: -901016616.
> 
> If you're working on a screen component, check your component configuration for invalid values.
>
> `This page has an error. You might just need to refresh it. [LWC component's @wire target property or method threw an error during value provisioning. Original error: [Cannot read property 'toLowerCase' of null]] Failing descriptor: {markup://c:fsc_flowCombobox}`

This occurs both when adding a new actions featuring a Flow Combobox in its CPE to the canvas, and when editing existing actions with a Flow Combobox. If a value like `{!VarOppoprtunity.Id}` exists in the Flow Combobx field already, then the error is immediately shown open opening the action editor.

The issue can be worked around if you paste the full path to the field you want in the variable, into the Flow Combobox field, e.g. `{!VarOpportunity.Name}`. Typing it manually causes the error to occur upon every keypress after the variable name, so pasting the path avoids this.

This issue was originally reported in issue #856, however this is not specific to the SendBetterEmail action, and is common to all actions and screen components using the fsc_flowCombobox component.

### Steps to reproduce

1. In a new screen flow, add a new record variable for any object, i.e. opportunity.
2. Add a new screen element to the canvas, and open it.
3. Add any custom component to the screen that uses a CPE, which features a Flow Comobox component, i.e. QuickChoice.
4. Click into the Flow Combobox field, i.e. _Value (Default or Existing)_, and click the **>** next to the record variable name.

### Expected behaviour

*   The record variable expands, and you can select a merge field from the object.

### Actual behaviour

*   The reported error occurs.

### Environment

I have reproduced this error on the following package versions:

- FlowScreenComponentsBasePack
  - 2.5.1
  - 2.5.3
- FlowActionsBasePack
  - 2.29
  - 2.31

### Screenshots

Here's the error occurring, as in the reproduction steps above:

![2021-08-12_22-26-05](https://user-images.githubusercontent.com/10392194/129272410-81d4f71b-d78b-444f-9ead-ab36465d15b1.gif)

**Comment by jdpond** [comment]
This evolved from #856 which was closed and moved to here.  For historical details, see #856

**Comment by ericrsmith35** [comment]
I did a simple test with QuickChoice and did not see this issue with either v2.5.0 or v2.5.4 of the FlowScreenComponentsBasePack.

**Comment by jdpond** [comment]
I just upgraded https://dream-force-2595-dev-ed.cs78.my.salesforce.com with:
```
ID                  Package ID          Package Name                  Namespace  Package Version ID  Version Name  Version
──────────────────  ──────────────────  ────────────────────────────  ─────────  ──────────────────  ────────────  ────────
0A31k000000AavVCAS  0334W000000hBcGQAU  FlowActionsBasePack           usf        04t4W0000034KebQAE  ver 0.1       2.32.0.0
0A31k000000AavkCAC  033f4000000mjCrAAI  SendBetterEmail                          04t5G000003rUZTQA2  ver 2.0.6     2.1.6.0
0A31k000000AavfCAC  033f4000000mjNrAAI  FlowScreenComponentsBasePack             04t5G000003rUaHQAU  ver 0.1       2.5.4.0
```
![image](https://user-images.githubusercontent.com/271131/129449734-0e92f574-3ce2-4665-bdf1-8bd92ba1747f.png)
Any ideas?

**Comment by ericrsmith35** [comment]
In your org, add some console.log statements in fsc_flowCombobox.js in the
getTypeOption and processOptions sections to track down what value ends up
being unexpectedly null.  Press CTRL-SHIFT-I to see the debug console while
you are running the CPE.  Those are the only 2 places in the combobox code
where the toLowerCase modifier is used.

**Comment by Garywoo** [comment]
@ericrsmith35 when you are trying to replicate this error, are you clicking on the variable name, or on the > button? The error only occurs when clicking the > button, which should then list the objects merge fields. 

![](https://user-images.githubusercontent.com/10392194/129267173-dc7ea587-1cfb-4326-8de1-9c1790f555ae.png)

**Comment by ericrsmith35** [comment]
I was clicking on the > button and then selecting a field name.

**Comment by jdpond** [comment]
@ericrsmith35 is this any more help?
```
/assets/icons/utility-sprite/svg/symbols.svg#check:1 Failed to load 
Failed to load resource: the server responded with a status of 404 (Not Found)
```
[Uncaught (in promise) TypeError: LWC component's @wire target property or method threw an error during value provisioning. Original error:
[Cannot read property 'toLowerCase' of null]
    at eval (fsc_flowCombobox.js:4)
    at Proxy.filter (<anonymous>)
    at eval (fsc_flowCombobox.js:4)
    at Proxy.forEach (<anonymous>)
    at _.processOptions (fsc_flowCombobox.js:4)
    at _.setOptions (fsc_flowCombobox.js:4)
    at _getObjectInfo (fsc_flowCombobox.js:4)
    at aura_prod.js:12
    at wi (aura_prod.js:12)
    at d.callback (aura_prod.js:12)](url)

**Comment by jdpond** [comment]
@ericrsmith35 
OK, I think I found the problem.   when the expand (>) is pressed, the localOptions instance (opToFilter) has the following values:
```
type: Opportunity Fields
Options: [collection of opportunity fields]
```
However, in the following line, "opToFilter.**displayType**.toLowerCase()" does not exist:
`localOptions = localOptions.filter(opToFilter => opToFilter.displayType.toLowerCase() === this.builderContextFilterType.toLowerCase() || opToFilter.storeOutputAutomatically === true || (  opToFilter.type === 'SObject' && !this.builderContextFilterCollectionBoolean));`

**Comment by ericrsmith35** [comment]
It looks like API 52.0 introduced a CloneSourceId reference in an SObject with a Type of null.  The fix I submitted in PR#870 avoids a CPE erroring out when trying to enter an Object>Field reference using fsc_flowCombobox.

**Comment by jdpond** [comment]
Udaman!  Thanks @ericrsmith35
